### ISSUE TITLE
FileService event reduction can cause exceptions in async events

### ISSUE THREAD

**Issue by Therzok** [mono/monodevelop#7276]
```
ERROR [2019-02-28 23:20:50Z]: An unhandled exception has occurred. Terminating Visual Studio? False
System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
  at System.Collections.Generic.List`1+Enumerator[T].MoveNextRare () [0x00013] in <3c73e0665b1b47bb9d553430a97aa942>:0
  at System.Collections.Generic.List`1+Enumerator[T].MoveNext () [0x0004a] in <3c73e0665b1b47bb9d553430a97aa942>:0
  at MonoDevelop.Projects.Project+<>c__DisplayClass349_0.<OnFileDeleted>b__0 () [0x00031] in <500e8d32b687408f897daf1edbb52aa9>:0
  at MonoDevelop.Core.Runtime+<>c.<RunInMainThread>b__40_0 (System.Object state) [0x00015] in <500e8d32b687408f897daf1edbb52aa9>:0
```

> VS bug [#806669](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/806669)

**Comment by mrward** [comment]
Kirill suggested using a .ToArray() here? Is that the solution? What is changing the collection after the event is fired?

**Comment by Therzok** [comment]
@mrward we're using the same FileCopyEventArgs for both AsyncEvents and non-Async.

AsyncEvents do not take into account whether the FileService is frozen, so we're sending those while processing _the same_ FileCopyEventArgs in EventQueue.

We need to consider either delaying AsyncEvents and sending them on Thaw() or cloning the eventargs.

**Comment by mrward** [comment]
Delaying the async events to be raised after a Thaw would change the behaviour of the existing implementation - at least based on the comment here - https://github.com/mono/monodevelop/blob/d111656014af56bd84ef0417c3bc0ff33493a450/main/src/core/MonoDevelop.Core/MonoDevelop.Projects/Project.cs#L4400-L4401

Cloning the event args is not ideal either I guess :(

**Comment by Therzok** [comment]
We can only clone during a freeze/thaw, that way the impact on the user experience is small (gc's will happen when the app is mostly in background).

**Comment by xamarin-release-manager** [comment]
> [sync] [VS-11] Comment by **xamarinc**



Fixed in version 8.1.0.103 (master)



Author: Marius Ungureanu  
 Commit: 34d54dd64ba022deb72ab8f5cf674f3170fcf1b6 (mono/monodevelop)

**Comment by xamarin-release-manager** [comment]
> [sync] [VS-14] Comment by **xamarinc**



Fixed in version 8.0.0.2709 (release-8.0)



Pull Request #7305 merged by: Lluis Sanchez  
 Author: mono  
 Commit: 72efd06d586dbe36d20d1b88e53a290bebc5b290 (xamarin/monodevelop)

**Comment by xamarin-release-manager** [comment]
> [sync] [VS-16] Comment by **xamarinc**



Fixed in version 8.1.0.1701 (master-vnext)



Pull Request #7305 merged by: Lluis Sanchez  
 Author: mono  
 Commit: 72efd06d586dbe36d20d1b88e53a290bebc5b290 (xamarin/monodevelop)

**Comment by xamarin-release-manager** [comment]
> [sync] [VS-24] Comment by **xamarinc**



Fixed in version 8.6.0.802 (master)



Pull Request #7305 merged by: Lluis Sanchez  
 Author: mono  
 Commit: 72efd06d586dbe36d20d1b88e53a290bebc5b290 (xamarin/monodevelop)
